Add "No sliding" mod for osu!

Feodor0090 opened the pull request:
Moved stream conversion code out of editor to utils class and added a mod that applies it.

Name of the mod is copy-pasted from mania one (since they do the same thing), maybe needs a rename.

Feodor0090 committed:
Move stream conversion code to utils class

Feodor0090 committed:
Implement Hold Off mod

Feodor0090 committed:
Add customization

Feodor0090 committed:
Add test scene

mk56-spn commented:
i was sceptical but this is actually ridiculously fun, that aside  personally i think it would be nice if kick sliders could be disabled to avoid overlapping horrible streams, other than that not much to say

peppy committed:
Merge branch 'master' into holds-off-std

peppy commented:
Agree with feedback on "kick sliders". Not sure about new combo, since that'll make some maps play a bit weirdly. Potentially it should add new combo if the combo gets too high, but not on every slider?

mk56-spn commented:
> Agree with feedback on "kick sliders". Not sure about new combo, since that'll make some maps play a bit weirdly. Potentially it should add new combo if the combo gets too high, but not on every slider?

- hmm, well my reasoning is doing the New combo thing feels better on average but i understand that it does make some maps feel a bit off. the proposal of making it do its thing if the combo goes too high seems sane but what cut off point can be used for that not to seem arbitrary?

further playtesting has brought a few minor issues to light on my end, notably lots of the artistic final stream that often go on forever, the issue being is that you can feasibly be able to do an entire map except the last slider because of stamina, which makes me think a setting slider which allows you to choose minimum slider duration before conversion and maximum, this nips the bud off annoying triple spam and  long end streams in the bud should the player decide they dont want to deal with it

Feodor0090 committed:
Add special convertor for kicksliders

Feodor0090 committed:
Allow nerfing too long streams

Feodor0090 committed:
Fix description

Feodor0090 committed:
Fix math in notes count calculation

Feodor0090 committed:
Use temporary list instead of yield to make convertor debuggable again

Feodor0090 committed:
Fix last note in a stream spawning at slider's beginning

Feodor0090 commented:
I added a workaround for kicksliders and ability to nerf deathstreams from long slow sliders.
I'm against giving generated streams new combos because we should keep flow of original map. If mapper wanted to connect a note with a slider with followpoints, then our note and stream must be connected too.

There is one detail that i want to explain before it will be asked, about `fe243ba6`.
While dealing with kicksliders i quite quickly stumbled upon #16990.
As mentioned in inline comment, `positionWithRepeats` is limited in [0, SpanCount], so for non-repeating slider it may be 1. Then, `pathPosition` will contain zero (because `1-(int)1` =0). I think that it was supposed, because `pathPosition = 1 - pathPosition` should reverse path position, setting it to something close to 1. But the check was not triggered because `positionWithRepeats%2` is exactly `1`, not something bigger than it.

Fixed via replacing comparison with precision by `>=`, so zero pathPosition will be correctly flipped.

Example (left is this branch, right is `2022.902.1`):

https://user-images.githubusercontent.com/53872073/188339519-1a9ebe9b-4333-4ee6-a17b-392f40b75f1b.mp4

Map is https://osu.ppy.sh/beatmapsets/1194661#osu/2490692

This maybe needs a test coverage, but i'm not sure how to write it because slider on picture below for me converts correctly on `2022.902.1` and is affected by this bug when running local build, so floating point errors also doing something here.


![Снимок экрана_2022-09-05_02-58-48](https://user-images.githubusercontent.com/53872073/188339590-cf18effa-2364-4fe9-8d3a-9bc859c8dfe6.png)

mk56-spn commented on osu.Game.Rulesets.Osu/Mods/OsuModHoldOff.cs:
```diff
@@ -0,0 +1,81 @@
+            MaxValue = 16
+        };
+
+        [SettingSource("Maximum stream length", "Streams longer than this value will be lightened.")]
```
you should probably add something like " length in beat measures" or "duration in beats " because most people wont know what you mean, or assume the sliders are being capped in actual on screen length

peppy commented:
As an aside, the name of this mod makes little sense in osu! ruleset. I don't think people will understand what it means.

Feodor0090 commented:
The only thing that i could imagine is `Drag off`. I mean, in osu!mania hold notes are being _holded_, while sliders in osu! are being _dragged_. Is this name okay? Unfortunately, I don't have any better ideas.

Feodor0090 committed:
Reword length setting a little

peppy commented:
Something like "No sliding" maybe?

Feodor0090 committed:
Rename the mod

Feodor0090 committed:
Merge branch 'master' into holds-off-std

AadilRn commented:
"Sliderless mod" I think might be a nicer sounding name maybe?

mk56-spn commented:
well i did even more playtesting and i ran into an edge case among edge cases that makes me think we should cap the maximum BPM of streams generated , aswell as refine the system for short repeat slider detection.

https://user-images.githubusercontent.com/74463310/190931258-39da8946-f13b-4426-bae1-15611d382f72.mov

as you can see yeah....3.3 stars => 8 

 dunno why the mapper decided this one diff needed  a 450 bpm change but they did, as such i suggest that in cases where the generated stream would surpass 350 bpm it should automatically get nerfed to half the bpm that the user set, lest waltz of the irish type maps end up being common place in ranked leaderboards
 
 and if im going to be honest the 16 x option should be removed its far too much.

A small aside, for long repeat slider there should be a small offset to the noted on each return so unpleasant not mapper intended perfect stacks arent created  

![osu_2022-09-19_00-47-59](https://user-images.githubusercontent.com/74463310/190931484-c1da856d-2c06-484a-948d-5a71419816c6.jpg)

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModNoSliding.cs:
```diff
@@ -0,0 +1,81 @@
+        };
+
+        [SettingSource("Maximum stream length in beats", "Streams with more notes than this value will be lightened.")]
+        public BindableInt MaxStreamLength { get; } = new BindableInt(16)
```
Not a huge fan of this setting. It's so arbitrary and needs knowledge of the map / map's timing.

Probably just remove it.

Feodor0090 commented:
@mk56-spn please link the map from the video.

> we should cap the maximum BPM of streams generated

Hard to disagree, if we are talking about 300-400-500 bpm. The problem here is that when choosing a map to play, player can see it's bpm and can customize the mod. It's bad if player set the mod to generate 1/4 streams but 1/2 (or even 1/1) were generated because the mod wants so. An option here is to nerf streams that have bpm much higher than other parts of a map. I need opinions about threshold. x2 from minimal bpm? From average? Limiting to specific values sounds bad for me.

Doxxz commented:
Game crashes on song select if you disable the mod while a map of unreasonable difficulty is selected.
https://osu.ppy.sh/beatmapsets/756794#osu/1605148

https://user-images.githubusercontent.com/67333710/191609292-45b621fa-12d3-41cd-8292-89ed4e0475dc.mp4

[runtime] 2022-09-21 20:52:05 [debug]: MouseDownEvent(Left) handled by SongSelect+SoloModSelectOverlay.
[runtime] 2022-09-21 20:52:05 [debug]: ClickEvent(Left) handled by IncompatibilityDisplayingModPanel.
[runtime] 2022-09-21 20:52:05 [debug]: MouseClick handled by IncompatibilityDisplayingModPanel.
[runtime] 2022-09-21 20:52:05 [error]: An unhandled error has occurred.
[runtime] 2022-09-21 20:52:05 [error]: System.InvalidOperationException: TransformStartTime at the end of delayed sequence is not the same as at the beginning, but should be. (begin=Infinity end=Infinity)
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Transforms.Transformable.<BeginDelayedSequence>b__32_0(DelayedSequenceSender sender)
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Allocation.ValueInvokeOnDisposal`1.Dispose()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Game.Graphics.UserInterface.StarCounter.animate(Single newValue) in /home/osu-holds-off-std/osu.Game/Graphics/UserInterface/StarCounter.cs:line 119
[runtime] 2022-09-21 20:52:05 [error]: at osu.Game.Graphics.UserInterface.StarCounter.set_Current(Single value) in /home/osu-holds-off-std/osu.Game/Graphics/UserInterface/StarCounter.cs:line 49
[runtime] 2022-09-21 20:52:05 [error]: at osu.Game.Screens.Select.Carousel.DrawableCarouselBeatmap.<ApplyState>b__34_0(ValueChangedEvent`1 d) in /home/osu-holds-off-std/osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:line 217
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Bindables.Bindable`1.TriggerValueChange(T previousValue, Bindable`1 source, Boolean propagateToBindings, Boolean bypassChecks)
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Bindables.Bindable`1.SetValue(T previousValue, T value, Boolean bypassChecks, Bindable`1 source)
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Bindables.Bindable`1.set_Value(T value)
[runtime] 2022-09-21 20:52:05 [error]: at osu.Game.Beatmaps.BeatmapDifficultyCache.<>c__DisplayClass29_1.<updateBindable>b__1() in /home/osu-holds-off-std/osu.Game/Beatmaps/BeatmapDifficultyCache.cs:line 218
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Threading.Scheduler.Update()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Platform.GameHost.UpdateFrame()
[runtime] 2022-09-21 20:52:05 [error]: at osu.Framework.Threading.GameThread.processFrame()
[runtime] 2022-09-21 20:52:05 [verbose]: Unhandled exception has been denied .
Unhandled exception. System.InvalidOperationException: TransformStartTime at the end of delayed sequence is not the same as at the beginning, but should be. (begin=Infinity end=Infinity)
   at osu.Framework.Graphics.Transforms.Transformable.<BeginDelayedSequence>b__32_0(DelayedSequenceSender sender)
   at osu.Framework.Allocation.ValueInvokeOnDisposal`1.Dispose()
   at osu.Game.Graphics.UserInterface.StarCounter.animate(Single newValue) in /home/osu-holds-off-std/osu.Game/Graphics/UserInterface/StarCounter.cs:line 119
   at osu.Game.Graphics.UserInterface.StarCounter.set_Current(Single value) in /home/osu-holds-off-std/osu.Game/Graphics/UserInterface/StarCounter.cs:line 49
   at osu.Game.Screens.Select.Carousel.DrawableCarouselBeatmap.<ApplyState>b__34_0(ValueChangedEvent`1 d) in /home/osu-holds-off-std/osu.Game/Screens/Select/Carousel/DrawableCarouselBeatmap.cs:line 217
   at osu.Framework.Bindables.Bindable`1.TriggerValueChange(T previousValue, Bindable`1 source, Boolean propagateToBindings, Boolean bypassChecks)
   at osu.Framework.Bindables.Bindable`1.SetValue(T previousValue, T value, Boolean bypassChecks, Bindable`1 source)
   at osu.Framework.Bindables.Bindable`1.set_Value(T value)
   at osu.Game.Beatmaps.BeatmapDifficultyCache.<>c__DisplayClass29_1.<updateBindable>b__1() in /home/osu-holds-off-std/osu.Game/Beatmaps/BeatmapDifficultyCache.cs:line 218
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Graphics.Drawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.processFrame()
--- End of stack trace from previous location ---
   at osu.Framework.Platform.GameHost.<>c__DisplayClass133_0.<abortExecutionFromException>b__0()
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Platform.ThreadRunner.RunMainLoop()
   at osu.Framework.Platform.GameHost.windowUpdate()
   at osu.Framework.Platform.SDL2DesktopWindow.Run()
   at osu.Framework.Platform.GameHost.Run(Game game)
   at osu.Desktop.Program.Main(String[] args) in /home/osu-holds-off-std/osu.Desktop/Program.cs:line 120

Feodor0090 committed:
Merge branch 'master' into holds-off-std

Feodor0090 committed:
Do not expose deathstream threshold

Feodor0090 committed:
Filter out circles at the same time (anti-2B)

Feodor0090 committed:
Reduce max beat divisor

Feodor0090 committed:
Slow down too fast beatmap parts

Feodor0090 commented:
I added guards against overlapping sliders and BPM peaks (decided to use 1.5 of minimal bpm as threshold), both maps in question are better now. Would appreciate further playtesting.

mk56-spn commented:
> I added guards against overlapping sliders and BPM peaks (decided to use 1.5 of minimal bpm as threshold), both maps in question are better now. Would appreciate further playtesting.

on a first pass it seems much better, not sure about having the guard against bpm be a toggle, i understand why you did it but i think it shouldnt be there atleast in the first iteration of the mod , to see if its something thats generally wanted. but thats a nitpick.

the overlapping slider issue still  need a bit of work imo. like making the path under slightly diverge forming a slight angle from where the repeat slider goes would solve the issue . but its not a massive issue regardless

Feodor0090 commented:
Test failure looks unrelevant.

> the overlapping slider issue still need a bit of work imo.

I have one idea how to completely solve it (slider rotation), but original idea was to just convert all "key holds" to "key spams" without changing original map's flow. So i will implement it only if people complain a lot about this.

Also, i'm not sure about anti-2b filtering performance, but that code has O(n-1) complexity, so maybe it's okay... Checking existing circles in the main loop will be much slower.

If there are no more edge cases, i'm requesting new review.

mk56-spn commented:
https://user-images.githubusercontent.com/74463310/193933291-77f63213-14eb-48c8-a037-93686bad90fa.mov

A little more playtesting and i caught this case, a converted repeat / kickslider can stack offscreen if its long enough , needs to be confined like random mod does

Feodor0090 commented:
The objects positions are ok. Stacking logic fails here. The same can be achieved by placing a circle in editor in top-left corner and stack a lot.

This definitely needs to be fixed. But i'm not sure, by changing stacking to confine objects or locally by generating some kind of pattern instead of placing notes at the same position.

mk56-spn commented:
> The objects positions are ok. Stacking logic fails here. The same can be achieved by placing a circle in editor in top-left corner and stack a lot.
> 
> This definitely needs to be fixed. But i'm not sure, by changing stacking to confine objects or locally by generating some kind of pattern instead of placing notes at the same position.

i suggest a pattern that basically is just following the kicksliders path ( but without going back on itself) , sure it might end up a bit tight but it avoids this

Feodor0090 commented:
To clarify, we are talking about kicksliders those head and tail are at the same positions, right? If head and tail are on different sides stacking logic is not triggering?

mk56-spn commented:
> To clarify, we are talking about kicksliders those head and tail are at the same positions, right? If head and tail are on different sides stacking logic is not triggering?

no i mean like, what you should do to avoid the stack leniency issue is basically place the first circle derived from the kickslider on the sliderhead position and the last one on the slider trail (or the one previous to that if the slider tail is on the same place as the head. if you dont get what ill make a little illustration

Feodor0090 committed:
Generate square if stack is going to be too long

Feodor0090 commented:
@mk56-spn I want your opinion about the thing i added.

mk56-spn commented:
> @mk56-spn I want your opinion about the thing i added. In theory this allows to avoid issues of stacks.

def helps, although personally id clamp the hitcircles from this pattern to the inner 95% of the playfield since they start stacking a bit out of the playfield if the kickslider is very long, other than that i think this is ready

might wanna adress the merge conflicts though   😅

Feodor0090 committed:
Count longer sliders as kicks too

Feodor0090 committed:
Clamp square to playfield

Feodor0090 commented:
> personally id clamp the hitcircles from this pattern to the inner 95% of the playfield

Agree. Did that.

> might wanna address the merge conflicts

It will become even more messy. IMO it's easier to reopen it, when it will considered ready for merge.
